packages/hebao_v1/contracts/base/MetaTxModule.sol

```diff
@@ -137,22 +156,21 @@ contract MetaTxModule is BaseModule
         address gasToken,
         uint    gasPrice,
         uint    gasLimit,
+        uint    gasOverhead,
         uint    gasSpent
         )
         private
     {
-        if (gasPrice == 0) return;
+        require(gasSpent <= gasLimit, "EXCEED_GAS_LIMIT");
```

Comment:
I would remove this check because the gas is now limited in the `call`, so if that call uses up all gas it still costs a bit of gas to call `reimburseGasFee` before this check is done, which could still fail the transaction. This would mean the relayer pays for the complete failing transactions even though he did everything right.

Reply:
If we remove this line, there there is no place we ever check against `gasLimit`.

Reply:
The gasLimit is used to measure the gas used since the first line of code in `executeMetaTx` but before we call `reimburseGasFee`:

```solidity
    function executeMetaTx(
        bytes   calldata data,
        uint    nonce,
        address gasToken,
        uint    gasPrice,
        uint    gasLimit,
        uint    gasOverhead,
        bytes[] calldata signatures
        )
        external
        payable
    {
        require(gasLimit > 0, "INVALID_GAS_LIMIT");

        uint startGas = gasleft();
        require(startGas >= gasLimit, "OUT_OF_GAS");

         [code measured by gasLimit]

        reimburseGasFee(
            wallet,
            gasToken,
            gasPrice,
            gasLimit,
            gasOverhead,
            startGas - gasleft()
        );
    }

```

The gasOverhead includes both the storage of calldata and the `reimburseGasFee` method.

Reply:
> If we remove this line, there there is no place we ever check against `gasLimit`.

We now do `address(this).call.gas(gasLimit.sub(startGas.sub(gasleft()))).value(msg.value)(data)` which makes it impossible to use more than `gasLimit` until after the call ended. But a bit more gas is used  for emitting the `ExecutedMetaTx` event and preparing the stack to call `reimburseGasFee` that is also measured until `startGas - gasleft()`. So it's possible `gasSpent` is slightly higher than `gasLimit`, even though this is completely valid (the call may fail having using up all requested gas and the relayer still gets paid, but the user is ultimately responsible to have a high enough `gasLimit`)

If we would do `require(gasSpent <= gasLimit, "EXCEED_GAS_LIMIT");` that would mean the transaction would always fail if the `call` uses it's complete amount of gas allowed of `gasLimit.sub(startGas.sub(gasleft()))`, which is certainly not what we want because the relayer would have simply wasted it's own gas for a failed transaction. It would be possible to let the `call` use a little bit less gas (e.g. subtract another 2,000 gas from `gaslimit`), but with changing gas prices for different operations on ETH 1.x I don't think that's a good idea.

Perhaps from a user point of view it makes more sense to just use `gasLimit` directly for the gas limit of the call itself (so `address(this).call.gas(gasLimit).value(msg.value)(data)`). Everything else (signature checking, emitting events, ...) can also be part of `gasOverhead`.  Users may know better how much the actual call costs (from having used Ethereum directly perhaps) and not really know how much the meta-tx part costs (as this will be different depending on the number of signatures etc... which the user may not really know, a different number of signatures may even be needed for doing the same operation in some cases). As `gasOverhead` is something that is automatically generated for users and `gasLimit` is something that users may be able to set themselves I think that would make sense.